### PR TITLE
feature CPD-399: Pytest API Unit Tests

### PR DESCRIPTION
**Pull request by izzy64**
#  #

## 🗣 Description ##

Adding pytest unit tests for api endpoints. While not exhaustive, the tests cover the basic GET method for most endpoints, and the method to get a specific object where applicable.

https://cset.atlassian.net/browse/CPD-399

## 💭 Motivation and context ##

The more tests we create, the better we can make sure our changes do not introduce bugs. 

## 🧪 Testing ##

The new tests pass with the current test data. 

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

- [ ] Revert dependencies to default branches.
- [ ] Finalize version.

## ✅ Post-merge checklist ##

- [ ] Add a tag or create a release.

### CODE REVIEW

File: tests/config_test.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+        resp = client.get("/api/config/")
+        assert resp.status_code == 200
+
+        self.check_config_properties(resp.json)
+
+    @staticmethod
+    def check_config_properties(config):
+        """Check config object for expected properties."""
+        try:
+            assert isinstance(config, dict)
```
Comment (xlyk) [inline tests/config_test.py]: are there any specific keys and values you can check for?

File: tests/logging_test.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    @staticmethod
+    def check_logging_properties(logging):
+        """Check logging object for expected properties."""
+        try:
+            assert isinstance(logging, list)
+        except KeyError:
+            pytest.fail("expected a list")
+
+        if len(logging) > 0:
```
Comment (xlyk) [inline tests/logging_test.py]: you can just use

```
if logging:
```

File: tests/logging_test.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+        self.check_logging_properties(resp.json)
+
+    @staticmethod
+    def check_logging_properties(logging):
+        """Check logging object for expected properties."""
+        try:
+            assert isinstance(logging, list)
+        except KeyError:
+            pytest.fail("expected a list")
```
Comment (xlyk) [inline tests/logging_test.py]: this won't work as you expect. I would do this

```
if not isinstance(logging, list):
    pytest.fail("expected a list")
```

there are a few spots in this pr where this is required

File: tests/nonhumans_test.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        """Check nonhuman object for expected properties."""
+        try:
+            assert isinstance(nonhuman[0], str)
+        except KeyError:
+            pytest.fail("expected a string")
+
+        try:
+            assert isinstance(nonhuman, list)
+            assert len(nonhuman) >= 3
+        except KeyError:
```
Comment (xlyk) [inline tests/nonhumans_test.py]: KeyError is only raised if a dict fails to find a key when you use a direct lookup

**Review by xlyk — CHANGES_REQUESTED** [review]

File: tests/logging_test.py (hunk 2 above)
Comment (izzy64) [reply tests/logging_test.py]: Ah, yes. Resolved with e3d9595830337df26d5c91a6890d6364d5eed8fe

File: tests/logging_test.py (hunk 3 above)
Comment (izzy64) [reply tests/logging_test.py]: Much better, thanks! Resolved with e3d9595830337df26d5c91a6890d6364d5eed8fe

File: tests/nonhumans_test.py (hunk 4 above)
Comment (izzy64) [reply tests/nonhumans_test.py]: Good point! Resolved with e3d9595830337df26d5c91a6890d6364d5eed8fe

File: tests/config_test.py (hunk 1 above)
Comment (izzy64) [reply tests/config_test.py]: Unfortunately, all of the values are defined as None during the github actions build, so the 'build / Test image (pull_request)' and 'build / Test image (push)' steps fail. See 1817fc51d583bc965afc1b7e3d77f8d5c77c33f7

image

Let me know if you know a way around this

Comment (izzy64) [reply tests/config_test.py]: oops forgot to @xlyk

File: tests/logging_test.py (hunk 2 above)
Comment (izzy64) [reply tests/logging_test.py]: oops forgot to @xlyk

File: tests/logging_test.py (hunk 3 above)
Comment (izzy64) [reply tests/logging_test.py]: oops forgot to @xlyk

File: tests/nonhumans_test.py (hunk 4 above)
Comment (izzy64) [reply tests/nonhumans_test.py]: oops forgot to @xlyk

**Review by xlyk — APPROVED** [review]

**Pull request merged**